deploy/src/crd.rs

```diff
@@ -0,0 +1,291 @@
+#![allow(dead_code)] // remove when it is implemented
+
+#[cfg(test)]
+use garde::Validate; // garde is for validation locally
+
+use k8s_openapi::api::core::v1::{PodResourceClaim, PodSecurityContext, Toleration};
+use k8s_openapi::serde::{Deserialize, Serialize};
+use kube::CustomResource;
+use schemars::JsonSchema;
+use std::collections::BTreeMap;
+
+/// Current CRD `XineCluster` version
+pub(crate) type Cluster = v1::Cluster;
+
+// The `CustomResource` macro generates a struct `Cluster` that does not pass the clippy lint.
+#[allow(clippy::str_to_string)]
+#[allow(clippy::missing_docs_in_private_items)]
+mod v1 {
+    #[cfg(test)]
+    use super::Validate;
+
+    use super::{
+        BTreeMap, CustomResource, Deserialize, JsonSchema, PodResourceClaim, PodSecurityContext,
+        Serialize, Toleration,
+    };
+
+    /// Xline cluster specification
+    #[derive(CustomResource, Deserialize, Serialize, Clone, Debug, JsonSchema)]
+    #[cfg_attr(test, derive(Validate))]
+    #[cfg_attr(test, garde(context(ClusterSpec)))]
+    #[kube(
+        group = "xlineoperator.datenlord.io",
+        version = "v1",
+        kind = "XlineCluster",
+        singular = "xlinecluster",
+        plural = "xlineclusters",
+        struct = "Cluster",
+        namespaced,
+        status = "ClusterStatus",
+        shortname = "xc",
+        scale = r#"{"specReplicasPath":".spec.size", "statusReplicasPath":".status.size"}"#,
+        printcolumn = r#"{"name":"CronSpec", "type":"string", "description":"The cron spec defining the interval a backup CronJob is run", "jsonPath":".spec.backup.cron"}"#,
+        printcolumn = r#"{"name":"Size", "type":"string", "description":"The cluster size", "jsonPath":".spec.size"}"#,
+        printcolumn = r#"{"name":"StorageType", "type":"string", "description":"The backup storage type", "jsonPath":".spec.backup.storage_type"}"#
+    )]
+    pub(crate) struct ClusterSpec {
+        /// Size of the xline cluster, less than 3 is not allowed
+        #[cfg_attr(test, garde(range(min = 3)))]
+        #[schemars(range(min = 3))]
```

Comment:
why this need two attributes

Reply:
The latter one is needed to ensure that a resource definition with a size smaller than 3 cannot be applied to k8s. However, we need an approach to validate the size in tests, which is what the former one does.